packages/ListBoxBrowser/src/helpers.js

```diff
@@ -0,0 +1,259 @@
+import React from "react";
+
+/** helpers */
+const hasOptions = options => typeof options !== "undefined" && Array.isArray(options);
+
+export function isRoot(key) {
+  if (key === null) return false;
+
+  return key.includes("root");
+}
+
+function getOptions(indexes, list) {
+  if (Array.isArray(indexes)) {
+    return indexes.map(index => {
+      return list[index].value;
+    });
+  }
+
+  return [list[indexes].value];
+}
+
+function isSelectable({ hasOptions, isParentSelectable }) {
+  if (isParentSelectable !== null) {
+    return !isParentSelectable;
+  }
+
+  return hasOptions;
+}
+
+/** DATA STRUCTURE */
+
+export function getData({
+  data,
+  path = "root",
+  refDefaultSelectedOptions,
+  defaultSelectedOptions,
+  defaultSelectedView,
+  refDefaultSelectedView,
+}) {
+  return data.map(({ label = null, options, ...moreAttributes }, index) => {
+    if (!label) {
+      throw new Error("A Label attribute is required for each option object, can't process data.");
+    }
+
+    let newPath = isRoot(path) ? `${index}` : `${path}/${index}`;
+    const recursiveParameters = {
+      data: options,
+      defaultSelectedOptions,
+      defaultSelectedView,
+      path: newPath,
+      refDefaultSelectedOptions,
+      refDefaultSelectedView,
+    };
+    const _hasOptions = hasOptions(options);
+
+    const option = {
+      parent: path,
+      hasOptions: _hasOptions,
+      $$key: newPath,
+      attributes: {
+        label,
+        options: _hasOptions ? getData(recursiveParameters) : null,
+        ...moreAttributes,
+      },
+    };
+
+    if (defaultSelectedOptions(option.attributes)) {
+      /* eslint-disable no-param-reassign */
+      if (isRoot(option.parent)) {
+        newPath = "root";
+      }
+
+      if (option.parent in refDefaultSelectedOptions.current) {
+        refDefaultSelectedOptions.current = {
+          ...refDefaultSelectedOptions.current,
+          [option.parent]: [...refDefaultSelectedOptions.current[option.parent], option],
+        };
+      } else {
+        refDefaultSelectedOptions.current = { ...refDefaultSelectedOptions.current, [option.parent]: [option] };
+      }
+      /* eslint-enable no-param-reassing */
+    }
+
+    // checking if its initial view
+    if (_hasOptions && defaultSelectedView && defaultSelectedView(option.attributes)) {
+      refDefaultSelectedView.current = newPath;
+    }
+
+    return option;
+  });
+}
+
+export function getOptionByKey(data, path) {
+  let node = null;
+  function runner(data, path) {
+    return data.map(option => {
+      if (!option.$$key.includes("/") && (typeof data[path] !== "undefined" && option.$$key === path)) {
+        node = option;
+        throw new Error("option found");
```

Comment:
Something more descriptive for the error would be nice :)

Reply:
so map doesn't support the break statement, what I'm doing is a technique where you throw an error to stop the loop, so what is actually happening I'm just describing the option has been found stop looping and during the handling exception, I return the value.

```js
  try {
    runner(data, path);
    return [];
  } catch (e) {
    return node;
  }
```

I took the inspiration from this gist I saw time ago.

https://gist.github.com/sebmarkbage/2c7acb6210266045050632ea611aebee#file-infrastructure-js-L23

Reply:
A+ for inventiveness 😮 👏 but I'm absolutely bummed to say that this is overengineered 😭 everything here is achievable and arguably made clearer with a plain loop. We don't need `.map` here because its return value is never used. Could replace it with `.forEach` but then it only overcomplicates by not being able to stop the loop.

Reply:
I agree map, is no necessary, I removed it and replaced for foreach and keep it the same way, throwing an exception on a loop make it easier to implement the recursive function. :)